Playwright: refactor EditorToolbarComponent and EditorNavSidebarComponent out of GutenbergEditorPage.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR refactors two new components out of GutenbergEditorPage:
- EditorToolbarComponent
- EditorNavSidebarComponent

Handling the top toolbar interactions and the editor navigation sidebar respectively.

Key changes:
- extract methods related to toolbar and navigation sidebar out of GutenbergEditorPage.
- update existing methods in GutenbergEditorPage to point to the extracted classes.

#### Testing instructions

Ensure that all following builds are green:
  - [x] WPCOM E2E (desktop)
  - [x] WPCOM E2E (mobile)
  - [x] Calypso E2E (desktop)
  - [x] Calypso E2E (mobile)
  - [x] Pre-Release

Related to https://github.com/Automattic/wp-calypso/issues/60868.
Depends on https://github.com/Automattic/wp-calypso/pull/61431.

worldomonation committed:
Implement EditorNavSidebarComponent class.

worldomonation committed:
Remove EditorNavSidebarComponent related methods and selectors from GutenbergEditorPage class.

Use new class for spec files.

worldomonation committed:
Implement initial phase of EditorToolbarComponent.
- saveDraft
- clickPublish
- switchToDraft

worldomonation committed:
Add complete set of methods to the EditorToolbarComponent.
- remove deprecated methods and selectors from GutenbergEditorPage.
- remove unnecessary skip and if/else clauses in Editor: Basic Post flow.

Fix logic for `targetIsOpen`

Restore a click on the main editor before opening block inserter.

Take into account mobile viewport difference when closing the settings panel

dpasque commented on packages/calypso-e2e/src/lib/components/editor-toolbar-component.ts:
```diff
@@ -0,0 +1,228 @@
+	 * @param {string} selector Target selector.
+	 * @returns {Promise<boolean>} True if target is in an expanded state. False otherwise.
+	 */
+	async targetIsOpen( selector: string ): Promise< boolean > {
```
I love this, this is very clever!

Can we make this private, since it's just an internal helper?

worldomonation replied:
Fixed in 4122ed003d9348d80efb3ad926838b891533454b.

dpasque commented on packages/calypso-e2e/src/lib/components/editor-nav-sidebar-component.ts:
```diff
@@ -0,0 +1,83 @@
+import { NavbarComponent } from '.';
+
+const panel = 'div.wpcom-block-editor-nav-sidebar-nav-sidebar__container';
+const selectors = {
+	sidebarButton: `button.wpcom-block-editor-nav-sidebar-toggle-sidebar-button__button`,
+	exitLink: `${ panel } a.wpcom-block-editor-nav-sidebar-nav-sidebar__home-button`,
+};
```
[Following up on my comment here](https://github.com/Automattic/wp-calypso/pull/61431#discussion_r816941921)... I think my personal preference for now would be to use text/aria based selectors for now until we need to explicitly support I18N. Totally optional, and if you disagree, these work just great! But wanted to tie the two comments together! 👍

worldomonation replied:
I'll give another day or so to see if the P2 post gets any traction - but if not, I will revert the selectors to use `aria-label` for some selectors that are much better off using those.

worldomonation replied:
I've changed some of the selectors to UI based ones in 3a888d682b0c6f9550c28c931959821d61d11cdb.
It also required some rework of the selector logic.

dpasque commented on packages/calypso-e2e/src/lib/components/editor-nav-sidebar-component.ts:
```diff
@@ -0,0 +1,83 @@
+		] );
+		const actions: Promise< unknown >[] = [ navigationPromise ];
+
+		if ( envVariables.VIEWPORT_NAME === 'mobile' ) {
+			// Mobile viewports do not use a sidebar.
+			const navbarComponent = new NavbarComponent( this.page );
+			actions.push( navbarComponent.clickMySites() );
```
I feel bad saying this because I know we're trying to take things out of the EditorPage, but this handling feels a little bit strange to me here. Per the single responsibility principle, I think we should try to make each component only responsible for itself, or components it "contains". So, the editor nav sidebar should only contain actions that are in that sidebar.

That practically in this case would mean pulling this logic into the EditorPage, which is a bummer, but I think that makes the most sense in terms of code hygiene!

What do you think?

worldomonation replied:
I've moved the logic up to the parent/calling method, `GutenbergEditorPage.exitEditor` in 4122ed003d9348d80efb3ad926838b891533454b.

dpasque commented on packages/calypso-e2e/src/lib/components/editor-toolbar-component.ts:
```diff
@@ -0,0 +1,228 @@
+const panel = 'div.interface-interface-skeleton__header';
+const selectors = {
+	// Block Inserter
+	blockInserterButton: `${ panel } button.edit-post-header-toolbar__inserter-toggle`,
+
+	// Draft
+	saveDraftButton: `${ panel } button.editor-post-save-draft[aria-disabled="false"]`,
+	saveDraftDisabledButton: `${ panel } button.editor-post-saved-state.is-saved[aria-disabled="true"]`,
+	switchToDraftButton: `${ panel } button.editor-post-switch-to-draft`,
+
+	// Preview
+	mobilePreviewButton: `${ panel } a.editor-post-preview`,
+	desktopPreviewButton: `${ panel } button.block-editor-post-preview__button-toggle`,
+	desktopPreviewMenuItem: ( target: PreviewOptions ) =>
+		`button[role="menuitem"] span:text("${ target }")`,
+	previewPane: ( target: PreviewOptions ) => `.is-${ target.toLowerCase() }-preview`,
+
+	// Publish
+	publishButton: `${ panel } button.editor-post-publish-button__button[aria-disabled="false"]`,
+
+	// Editor settings
+	settingsButton: `${ panel } .interface-pinned-items > button:first-child`,
```
Same question here re: text selectors

worldomonation replied:
As with https://github.com/Automattic/wp-calypso/pull/61502#discussion_r818250311, some selectors were changed with UI based ones where it made sense.

dpasque commented on packages/calypso-e2e/src/lib/components/editor-toolbar-component.ts:
```diff
@@ -0,0 +1,228 @@
+	/**
+	 * Opens the block inserter.
+	 */
+	async openBlockInserter(): Promise< void > {
+		if ( ! ( await this.targetIsOpen( selectors.blockInserterButton ) ) ) {
+			const locator = this.frameLocator.locator( selectors.blockInserterButton );
+			await locator.click();
+		}
+	}
+
+	/**
+	 * Closes the block inserter.
+	 */
+	async closeBlockInserter(): Promise< void > {
+		if ( await this.targetIsOpen( selectors.blockInserterButton ) ) {
+			const locator = this.frameLocator.locator( selectors.blockInserterButton );
+			await locator.click();
+		}
+	}
```
This is really nicely designed throughout this class 🙌  -- I love these juxtaposed methods that guarantee their ending state!! 🎉

dpasque commented on packages/calypso-e2e/src/lib/components/editor-toolbar-component.ts:
```diff
@@ -0,0 +1,228 @@
+	async saveDraft(): Promise< void > {
+		const saveButtonLocator = this.frameLocator.locator( selectors.saveDraftButton );
+
+		if ( ! saveButtonLocator.waitFor( { timeout: 5 * 1000 } ) ) {
```
I believe this throws when it times out, right? We might need a try/catch here instead.

worldomonation replied:
Good point. I will re-think how this part is structured, there _must_ be a better way to do this.

worldomonation replied:
I've wrapped the `waitFor` call around a `try/catch` for now.

I almost went back to using `locator.count`. My personal feeling is that `locator.count` is a reliable way to compute the number of elements that match a selector at any given time, as recall that's the point of `locator` API. The only caveat with `locator.count` is that if a page is not loaded fully, extrapolating from how behavior for other methods like `page.$$` have behaved we may not get the full picture, but in practice I have not observed this issue with `locator.count`.

worldomonation replied:
Fixed in 4122ed003d9348d80efb3ad926838b891533454b.

dpasque commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -351,56 +328,39 @@ export class GutenbergEditorPage {
 		}
-		const frame = await this.getEditorFrame();
-		await frame.click( selectors.closeSettingsButton );
+		await this.editorToolbarComponent.closeSettings();
```
I believe we're missing an `else` here? We don't need to click on any toolbar buttons in the mobile case, right?

worldomonation replied:
The else is not necessary here, since the `EditorToolbarComponent.closeSettings()` checks whether the `Settings` is open or not prior to running the `click` action. In this case, after the mobile viewport handling of `await editorSettingsSidebarComponent.closeSidebar()` the check for whether the setting sidebar is open returns a `false`, so the `closeSettings` method short circuits.

dpasque commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -518,137 +462,54 @@ export class GutenbergEditorPage {
+			return await this.editorToolbarComponent.openMobilePreview();
 		}
-		const frame = await this.getEditorFrame();
-		await frame.click( selectors.previewButton );
-		await frame.click( selectors.previewMenuItem( target ) );
-		const handle = await frame.waitForSelector( selectors.previewPane( target ) );
-		await handle.waitForElementState( 'stable' );
+		if ( ! target ) {
+			throw new Error( 'Preview target must be defined.' );
+		}
+
+		await this.editorToolbarComponent.openDesktopPreview( target );
 	}
```
Could you share a little more about wanting to collapse the two preview methods across viewports here?

In general, I'm definitely very much in favor of that (and actually just recently put this in the docs!), but I wonder if in this case it makes things cleaner to make an exception and keep them separate. Namely, I'm thinking because those two actions essentially required different method signatures! And collapsing those signatures here means that everyone who calls this method is already going to have to write a bunch of logical handling around the parameter values and differing return types. Given that, I think it might be clearer to just keep the two distinct methods.

What do you think?

worldomonation replied:
I wanted to keep things in line with rest of the refactored files where there exists _one_and only one obvious way to perform an action.

Your point about the caller having to write handlers for the `void|Page` return value is true and it just might be that there's no way around this other than splitting the preview methods into two. Perhaps the way preview is structured means there isn't a good way to unify the interface.

worldomonation replied:
Fixed in 9d1c1774ecf61e7a2878264695ec9d8fd0c14bf0.

dpasque reviewed (commented):
Great work overall! You did such a nice job with the design of the new components! 🎉 

Have a few kind of bigger picture questions, let me know what you think!

worldomonation committed:
Changes:
- rework the if/else conditions for single-purpose methods (eg. `closePanel`)
- implement new method in `GutenbergEditorPage` called `closeAllPanels`.

worldomonation committed:
Address review comments:
- move logic for EditorNavSidebarComponent.exitEditor out to the parent class.
- turn `targetIsOpen` method into private.

worldomonation committed:
Address review comments:
- split preview methods into two.

worldomonation committed:
Use UI based selectors.

dpasque commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -255,6 +248,32 @@ export class GutenbergEditorPage {
+	 * 	- Editor Navigation Sidebar
+	 */
+	async closeAllPanels(): Promise< void > {
+		try {
```
100% optional, but wanted to throw out the alternative syntax of
```
await Promise.allSettled([ // these three actions ])
```

worldomonation replied:
Learn something new every day. Thanks for the suggestion!

worldomonation replied:
Fixed in 781fadad87fb98107179c863cb5013ff141a6030.

dpasque commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -255,6 +248,32 @@ export class GutenbergEditorPage {
+		try {
+			await this.editorPublishPanelComponent.closePanel();
+		} catch {
+			// noop
```
Should we just let this throw if it errors? All `close` methods are designed to be error safe, so if something _does_ error here, I figure it might actually be meaningful and worth bubbling up!

worldomonation replied:
I've switched to use the `Promise.allSettled` construct so in that case, any uncaught error would be thrown, so this should be good?

worldomonation replied:
Fixed in 781fadad87fb98107179c863cb5013ff141a6030.

dpasque commented on packages/calypso-e2e/src/lib/components/editor-toolbar-component.ts:
```diff
@@ -0,0 +1,229 @@
+	/**
+	 * Opens the block inserter.
+	 */
+	async openBlockInserter(): Promise< void > {
+		if ( ! ( await this.targetIsOpen( selectors.blockInserterButton ) ) ) {
+			const locator = this.frameLocator.locator( selectors.blockInserterButton );
+			await locator.click();
+		}
+	}
+
+	/**
+	 * Closes the block inserter.
+	 */
+	async closeBlockInserter(): Promise< void > {
+		if ( await this.targetIsOpen( selectors.blockInserterButton ) ) {
+			const locator = this.frameLocator.locator( selectors.blockInserterButton );
+			await locator.click();
+		}
+	}
```
100% optional in this PR and we can do it later, but what do you think about adding this pattern to the docs? I was just thinking how much I love this design pattern you've set up where POM methods are essentially starting-state-safe, and the name of the method guarantees the ending state. I really want others to follow this! I think it makes the classes way easier to read and the scripts much more false-positive safe!

worldomonation replied:
Let's add it into the official style guide the following:
- method names should state desired end result, or the action performed in the method;
  - principle of "let this be done, and if not applicable, fail safely"
- for methods that have binary outcomes, create methods that handle each case, following the guide above;

dpasque reviewed (approved):
A few small optional suggestions, but everything looks great and ready to roll! 🚀 

This is a fantastic refactor -- I think the editor is in a **much** better place after this!! 🎉

worldomonation committed:
Use `Promise.allSettled` construct to close all panels

worldomonation committed:
Add a wait for the unpublish

worldomonation committed:
Fix selector for unpublish snackbar toast

worldomonation committed:
Allow state for `publishButton` selector.
